bitvm/src/hash/blake3.rs

```diff
@@ -235,416 +235,215 @@ pub fn blake3(
+        .into_iter()
+        .flat_map(|chunk| chunk.collect::<Vec<u8>>().into_iter().rev())
+        .chunks(64) // collect 64-byte chunks
+        .into_iter()
+        .map(|mut chunk| std::array::from_fn(|_| chunk.next().unwrap()))
+        .collect()
+}
 
-    assert!(padded_msg.len() % 64 == 0, "padding failed");
+/// Returns a script that pushes the given message onto the stack for BLAKE3.
+///
+/// The script transforms the message into the correct format
+/// and pushes the result onto the stack.
+///
+/// ## Panics
+///
+/// This function panics if the message is longer than 1024 bytes,
+/// since our BLAKE3 implementation doesn't support longer messages.
+pub fn blake3_push_message_script(message_bytes: &[u8]) -> Script {
+    assert!(
+        message_bytes.len() <= 1024,
+        "This BLAKE3 implementation doesn't support messages longer than 1024 bytes"
+    );
+    let chunks = chunk_message(message_bytes);
 
-       // reverse the 4 byte chunks
-    for chunk in padded_msg.chunks_mut(4) {
-        chunk.reverse();
-    }
- 
-    // push the msg into the stack and compact it
-    stack.custom(script!(
-        for chunk in padded_msg.chunks(64).rev(){
-            for (i,byte) in chunk.iter().enumerate(){
-                {*byte}
-                if i == 31 || i == 63{
-                    {U256::transform_limbsize(8,29)}
+    script! {
+        for chunk in chunks.into_iter().rev() {
+            for (i, byte) in chunk.into_iter().enumerate() {
+                {
+                    byte
+                }
+                if i == 31 || i == 63 {
+                    {
+                        U256::transform_limbsize(8, 29)
+                    }
                 }
             }
-        } 
-    ),
-    0,
-    false,
-    0,
-    "push_msgs"
-    );
+        }
+    }
+}
 
-    blake3(&mut stack, msg_len as u32, true, use_full_tables);
+/// Returns a script that computes the BLAKE3 hash of the message on the stack.
+///
+/// ## Panics
+///
+/// This function panics if the message is longer than 1024 bytes,
+/// since our BLAKE3 implementation doesn't support longer messages.
+pub fn blake3_compute_script(message_len: usize) -> Script {
+    assert!(
+        message_len <= 1024,
+        "This BLAKE3 implementation doesn't support messages longer than 1024 bytes"
+    );
+    let mut stack = StackTracker::new();
+    let use_full_tables = true;
+    blake3(&mut stack, message_len as u32, true, use_full_tables);
+    stack.get_script()
+}
 
-    //change the hash representation from nibbles to bytes and compare with expected hash value
-    stack.custom(script!(
-        for (i, byte) in expected_hash.as_bytes().iter().enumerate(){
-            {*byte}
-            if i % 32 == 31{
+/// Returns a script that verifies the BLAKE3 output on the stack.
+///
+/// The script pops the BLAKE3 output and compares it with the given, expected output.
+pub fn blake3_verify_output_script(expected_output: [u8; 32]) -> Script {
```

Comment:
Why not accept an array reference here? This would be more flexible for the caller.

Reply:
Blake3 hashes are always 32-byte values. There is no reason to accept slices.

Reply:
There is actually a movement in rust-bitcoin and rust-secp to replace slice references with arrays, when possible. It more clearly communicated which values are possible.

Reply:
Right, but you can type it `&[u8; 32]`, right? Then you can ensure the required length while giving the caller the freedom to avoid clones, perhaps.

Reply:
As discussed privately, types that implement `Copy` should be passed by value. A function that takes an owned value has more flexibility in what it can do with it. For example, the script that we construct in `blake3_verify_output_script` iterates over the expected output bytes. The version on `main` uses `*byte` to deference each byte. My version can directly use `byte` because it has an owned array.